cpp/src/arrow/dataset/scanner_internal.h

```diff
@@ -0,0 +1,87 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <utility>
+
+#include "arrow/dataset/dataset_internal.h"
+#include "arrow/dataset/filter.h"
+#include "arrow/dataset/scanner.h"
+
+namespace arrow {
+namespace dataset {
+
+static inline RecordBatchIterator FilterRecordBatch(
+    RecordBatchIterator it, std::shared_ptr<Expression> filter,
+    std::shared_ptr<ExpressionEvaluator> evaluator) {
+  if (filter == nullptr || evaluator == nullptr) {
```

Comment:
It was required in one of the unit test.

Reply:
I see. I think this is due to https://github.com/apache/arrow/pull/5675/files#diff-91c3d3ad9e9a52dd2886621a692bcf6eR85 being default constructed (instead of being set to `scalar(true)`). We should fix this in a follow up PR